File: examples/uhd_rx_compressed_cfile/fft_bin_calculator.py
#!/usr/bin/env python3

import sys
from math import ceil, floor

# For Pluto fft_size is 1024, capture_bw is 61.44, filter_bw is 56
fft_size   = 2048
half_fft   = int(fft_size/2)
capture_bw = 100
filter_bw  = 100 # set to 40 for WBX
rnd_digits = 4


def bins_calc (ccf, bcf, bw):
  # Compute frequency ranges
  binw = capture_bw/fft_size
  cl   = ccf - (capture_bw*1.0/2)
  fl   = ccf - (filter_bw*1.0/2)
  fr   = ccf + (filter_bw*1.0/2)
  bl   = bcf - (bw*1.0/2)
  br   = bcf + (bw*1.0/2)

  if (bl<fl) or (br>fr):
    print ("Band frequency out of filter range.")
    return
  # Compute FFT bin number, round to avoid python float errors
  bl = round((bl - cl)/binw, rnd_digits)
  br = round((br - cl)/binw, rnd_digits)

  # find the integer values for bin number
  bl = floor(bl)
  if (ceil(br)==int(br)):
    br = int(br) - 1
  else:
    br = floor(br)

  # FFT half window shift, so center frequency is at bin 0
  bl = bl ^ half_fft
  br = br ^ half_fft

  # check if the range is continuous or not
  if (bl<half_fft and br<half_fft) or (bl>(half_fft-1) and br>(half_fft-1)):
    print (bl, br)
  elif (bl==half_fft and br==(half_fft-1)):
    print (0, fft_size-1)
  else:
    print (0, br, "and", bl, fft_size-1)


if __name__ == "__main__":
  if (len(sys.argv)<4):
    print ("Missing arguments: <capture center freq> <band center freq> <bandwidth>")
  else:
    bins_calc(float(sys.argv[1]), float(sys.argv[2]), float(sys.argv[3]))

File: utilities/fft_bin_calculator.c
#include <stdlib.h>
#include <stdio.h>
#include <math.h>

const int rnd_digits = 4;

struct exact_ranges {
		float        l_freq;
		float        r_freq;
		unsigned int l_bin1;
		unsigned int r_bin1;
		unsigned int l_bin2;
		unsigned int r_bin2;
};

float round_float (float val, int rnd_digits){
    return (round(val * pow(10, rnd_digits)) / pow(10, rnd_digits));
}

int bins_calc (float capture_center_freq, float capture_bw,
								float band_center_freq, float band_bandwidth,
								float filter_bw, unsigned int fft_size,
								struct exact_ranges * final_ranges){

	  int   half_fft, left_bin, right_bin;
    float bin_width, capture_left, filter_left, filter_right,
		      band_left, band_right;

    // Compute frequency ranges
    bin_width    = capture_bw / fft_size;
    capture_left = capture_center_freq - (capture_bw     /2);
    filter_left  = capture_center_freq - (filter_bw      /2);
    filter_right = capture_center_freq + (filter_bw      /2);
    band_left    = band_center_freq    - (band_bandwidth /2);
    band_right   = band_center_freq    + (band_bandwidth /2);

    // Frequency range check
	  if ((band_left<filter_left) || (band_right>filter_right)){
        return 0;
	  }

    // Compute FFT bin number, round to avoid float errors
    band_left  = round_float((band_left  - capture_left) / bin_width, rnd_digits);
    band_right = round_float((band_right - capture_left) / bin_width, rnd_digits);

    // find the integer values for bin number
    left_bin = floor(band_left);
    if (ceil(band_right)==floor(band_right))
        right_bin = floor(band_right) - 1;
    else
        right_bin = floor(band_right);

    // Frequency range captured
    final_ranges->l_freq = round_float(capture_left + left_bin     *bin_width, rnd_digits);
    final_ranges->r_freq = round_float(capture_left + (right_bin+1)*bin_width, rnd_digits);

    // FFT half window shift, so center frequency is at bin 0
	  half_fft  = fft_size/2;
    left_bin  = left_bin ^ half_fft;
    right_bin = right_bin ^ half_fft;

    // check if the range is continuous or not
    if (((left_bin < half_fft    ) && (right_bin < half_fft    )) ||
	      ((left_bin > (half_fft-1)) && (right_bin > (half_fft-1)))){

				final_ranges-> l_bin1 = left_bin;
				final_ranges-> r_bin1 = right_bin;
				final_ranges-> l_bin2 = 0;
				final_ranges-> r_bin2 = 0;
				return 1;

    } else if ((left_bin == half_fft) && (right_bin == (half_fft-1))) {

				final_ranges-> l_bin1 = 0;
				final_ranges-> r_bin1 = fft_size-1;
				final_ranges-> l_bin2 = 0;
				final_ranges-> r_bin2 = 0;
				return 1;

    } else {

				final_ranges-> l_bin1 = 0;
				final_ranges-> r_bin1 = right_bin;
				final_ranges-> l_bin2 = left_bin;
				final_ranges-> r_bin2 = fft_size-1;
				return 2;

		}
}

int main( int argc, char *argv[] ) {
		struct exact_ranges ranges;
		int    range_count;

	  if (argc!=4) {
        printf("Missing arguments: <capture center freq> <band center freq> <band bandwidth>\n");
	  } else {
				// Pluto:
				// range_count = bins_calc(atof(argv[1]), 61.44, atof(argv[2]), atof(argv[3]), 56.0, 1024, ranges);
				// N210+WBX:
				// range_count = bins_calc(atof(argv[1]), 100.0, atof(argv[2]), atof(argv[3]), 40.0, 2048, ranges);
				// N210+SBX:
				range_count = bins_calc(atof(argv[1]), 100.0, atof(argv[2]), atof(argv[3]), 100.0, 2048, &ranges);

				if (range_count == 0){
						printf ("Band frequency out of filter range.\n");
				} else if (range_count == 1) {
						printf ("Frequency range: %.*f to %.*f\n", rnd_digits, ranges.l_freq, rnd_digits, ranges.r_freq);
						printf ("FFT range      : %d to %d\n", ranges.l_bin1, ranges.r_bin1);
				} else {
						printf ("Frequency range: %.*f to %.*f\n", rnd_digits, ranges.l_freq, rnd_digits, ranges.r_freq);
						printf ("FFT ranges     : %d to %d and %d to %d\n", ranges.l_bin1, ranges.r_bin1, ranges.l_bin2, ranges.r_bin2);
				}
		}

	  return 0;
}
